src/examples/zkapps/voting/test.ts

```diff
@@ -31,43 +42,640 @@ export async function testSet(
+    );
+  }
+
+  /*
+    test case description:
+      registering candidate within the election period
+    
+    preconditions:
+      - slot has been set to within the election period
+
+    tested cases:
+      - registering candidate -> failure
+      - registering voter -> failure
+  
+    expected results:
+      - no new events emitted
+      - no state changes
+  */
+
+  console.log(
+    'attempting to register a candidate within the election period ...'
+  );
+  Local.setGlobalSlot(params.electionPreconditions.startElection.add(1));
+
+  let previousEventsVoter = voterContract.reducer.getActions({}).length;
+  let previousEventsCandidate = candidateContract.reducer.getActions({}).length;
+
+  await assertValidTx(
+    false,
+    () => {
+      let lateCandidate = Member.from(
+        PrivateKey.random().toPublicKey(),
+        Field.zero,
+        UInt64.from(200)
+      );
+      // register late candidate
+      voting.candidateRegistration(lateCandidate);
+      voting.sign(votingKey);
+    },
+    feePayer,
+    'rangeCheckHelper'
+  );
+
+  console.log('attempting to register a voter within the election period ...');
+
+  await assertValidTx(
+    false,
+    () => {
+      let lateVoter = Member.from(
+        PrivateKey.random().toPublicKey(),
+        Field.zero,
+        UInt64.from(50)
+      );
+
+      // register late voter
+      voting.voterRegistration(lateVoter);
+      voting.sign(votingKey);
+    },
+    feePayer,
+    'rangeCheckHelper'
+  );
+
+  if (previousEventsVoter !== voterContract.reducer.getActions({}).length) {
+    throw Error('events emitted but should not have been');
+  }
+  if (
+    previousEventsCandidate !== candidateContract.reducer.getActions({}).length
+  ) {
+    throw Error('events emitted but should not have been');
+  }
+
+  if (
+    !candidateContract.committedMembers
+      .get()
+      .equals(candidatesStore.getRoot())
+      .toBoolean()
+  ) {
+    throw Error(
+      'candidatesStore merkle root does not match on-chain committed members'
     );
-    candidateContract.accumulatedMembers.set(
-      Experimental.Reducer.initialActionsHash
+  }
+
+  if (
+    !voterContract.committedMembers
+      .get()
+      .equals(votersStore.getRoot())
+      .toBoolean()
+  ) {
+    throw Error(
+      'votersStore merkle root does not match on-chain committed members'
+    );
+  }
+
+  /*
+    test case description:
+      attempting to count votes before any votes were casted
+    
+    preconditions:
+      - no votes have been casted
+
+    tested cases:
+      - count votes -> success, but no state change
+
+    expected results:
+      - no state change
+  */
+  console.log('attempting to count votes but no votes were casted...');
+
+  let beforeAccumulator = voting.accumulatedVotes.get();
+  let beforeCommitted = voting.committedVotes.get();
+  await assertValidTx(
+    true,
+    () => {
+      voting.countVotes();
```

Comment:
It will be good to print results as part of the `countVotes` method invocation rather then as part of another helper method `printResult`. Wdyt?

Reply:
unfortunately not, that would be a pretty hacky way since we dont have access to the actual vote-data inside the contract, only a reference to it (the committed storage root) :/